sms-mms-import-to-signal.py

```diff
@@ -17,173 +35,229 @@ def get_contacts(cursor):
             contacts_by_number[c[1].replace("+61", "0").replace("-", "")] = c[0]
     return contacts_by_number
 
+def get_groups(cursor):
+    cursor.execute("select _id, group_id, recipient_id, members from group")
+    groups = cursor.fetchall()
+    groups_by_number = {}
+    for g in groups:
+        groups_by_number['_id'] = g[0]
+        groups_by_number['group_id'] = g[1]
+        groups_by_number['recipient_id'] = g[2]
+        groups_by_number['members'] = g[3]
+    return groups_by_number
 
 def get_parts(r):
     rtn = []
-    # print("GETTING PARTS:")
+    part_length = 0
+    logging.debug("GETTING PARTS:")
     for parts in r.findall("parts"):
         for part in parts.findall("part"):
+            if int(part.attrib.get("seq", 0)) != -1 and part.attrib.get("data"):
+                part_length += 1
             rtn.append(part)
-    return rtn
+    return part_length, rtn
 
 
 def get_addrs(r):
+    #this is a list of addresses, type 151 is you, type 130 are other recipients, type 137 is the sender. this will be useful for groups (need to figure how signal marks the sender in a group) also usefu
+    #this returned list can be used to allocate messages to the sender of a group
     rtn = []
-    # print("GETTING ADDRS:")
+    logging.debug("GETTING ADDRS:")
     for addrs in r.findall("addrs"):
         for addr in addrs.findall("addr"):
             rtn.append(addr)
     return rtn
 
+def add_recipient(add, cursor):
+    global contacts_by_number
+    cursor.execute(f"""insert into recipient (phone, default_subscription_id, registered) values ("{add}", 1, 2)""")
+    conn.commit()
+    contacts_by_number = get_contacts(cursor)
+    contacts_by_number[add]
```

Comment:
This line isn't doing anything

Reply:
that is right, late night 'write until it works' sessions 🙂